File: overlay/tabletop_object_perception/tabletop_collision_map_processing/src/tabletop_collision_map_processing/collision_map_interface.py
#! /usr/bin/python
# Software License Agreement (BSD License)
#
# Copyright (c) 2009, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# author: Kaijen Hsiao

## @package collision_map_interface
# Python version of collision_map_interface.cpp
# Functions that take and reset the collision map, and add, remove, attach
# or detach objects

import roslib; roslib.load_manifest('tabletop_collision_map_processing')
import random, time, math, scipy, pdb
import rospy
from geometric_shapes_msgs.msg import Shape
from tabletop_object_detector.msg import Table
from collision_environment_msgs.msg import MakeStaticCollisionMapAction, MakeStaticCollisionMapGoal
from mapping_msgs.msg import CollisionObject, AttachedCollisionObject, CollisionObjectOperation
from actionlib_msgs.msg import GoalStatus
from std_srvs.srv import Empty, EmptyRequest
from geometric_shapes_msgs.msg import Shape
import actionlib
import tf
from geometry_msgs.msg import PoseStamped, Point, Quaternion, Pose

#interface to the collision map
class CollisionMapInterface():

    def __init__(self):

        #set up service clients, publishers, and action clients
        rospy.loginfo("waiting for collision_map_self_occ_node/reset service")
        rospy.wait_for_service("collision_map_self_occ_node/reset")
        self.collision_map_reset_client = rospy.ServiceProxy("collision_map_self_occ_node/reset", Empty)
        
        self.make_static_collision_map_client = actionlib.SimpleActionClient('make_static_collision_map', MakeStaticCollisionMapAction)
        rospy.loginfo("waiting for make_static_collision_map action server")
        self.make_static_collision_map_client.wait_for_server()
        
        rospy.loginfo("advertising on collision object topics")
        self.object_in_map_pub = rospy.Publisher("collision_object", CollisionObject)
        self.attached_object_pub = rospy.Publisher("attached_collision_object", AttachedCollisionObject)

        self.object_in_map_current_id = 0
        rospy.loginfo("done initializing collision map interface")


    ##reset the current collision map to default
    def reset_collision_map(self):

        #reset the collision matrix
        try:
            resp = self.collision_map_reset_client(EmptyRequest())
        except rospy.ServiceException, e:
            rospy.logerr("error when calling collision_map_self_occ_node/reset: %s"%e)
            return 0

        #remove all objects
        reset_object = CollisionObject()
        reset_object.operation.operation = CollisionObjectOperation.REMOVE
        reset_object.header.frame_id = "base_link"
        reset_object.header.stamp = rospy.Time.now()
        reset_object.id = "all"
        self.object_in_map_pub.publish(reset_object)

        #and all attached objects
        reset_attached_objects = AttachedCollisionObject()
        reset_attached_objects.link_name = "all"
        reset_attached_objects.object.header.frame_id = "base_link"
        reset_attached_objects.object.header.stamp = rospy.Time.now()
        reset_attached_objects.object = reset_object
        self.attached_object_pub.publish(reset_attached_objects)

        rospy.loginfo("collision map reset")
        self.object_in_map_current_id = 0.
        return 1


    ##take a static map scan
    def take_static_map(self):
        static_map_goal = MakeStaticCollisionMapGoal()
        static_map_goal.cloud_source = "full_cloud_filtered"
        static_map_goal.number_of_clouds = 2;
        self.make_static_collision_map_client.send_goal(static_map_goal)

        if not self.make_static_collision_map_client.wait_for_result(rospy.Duration(30)):
            rospy.loginfo("collision map was not formed in allowed time")
            return 0

        if self.make_static_collision_map_client.get_state() == GoalStatus.SUCCEEDED:
            rospy.loginfo("static map successfully updated")
            return 1
        else:
            rospy.loginfo("some other non-success state was reached for static collision map.  Proceed with caution.")
            return 0


    ##attaches an object to the gripper for the purposes of collision-aware 
    #motion planning
    def attach_object_to_gripper(self, object_name, whicharm = 'r'):

        obj = AttachedCollisionObject()
        
        obj.link_name = whicharm+"_gripper_r_finger_tip_link"
        obj.object.operation.operation = CollisionObjectOperation.ATTACH_AND_REMOVE_AS_OBJECT
        obj.object.header.stamp = rospy.Time.now()
        obj.object.header.frame_id = "base_link"
        obj.object.id = object_name
        touch_links = ["_gripper_palm_link", "_gripper_r_finger_tip_link", "_gripper_l_finger_tip_link", "_gripper_l_finger_link", "_gripper_r_finger_link"]
        obj.touch_links = [whicharm+link for link in touch_links]
        self.attached_object_pub.publish(obj)


    ##detaches all objects from the gripper 
    #(removes from collision space entirely)
    def detach_all_objects_from_gripper(self, whicharm = 'r'):

        rospy.loginfo("detaching all objects from gripper %s"%whicharm)
        obj = AttachedCollisionObject()
        obj.object.header.stamp = rospy.Time.now()
        obj.object.header.frame_id = "base_link"
        obj.link_name = whicharm+"_gripper_r_finger_tip_link"
        obj.object.operation.operation = CollisionObjectOperation.REMOVE
        self.attached_object_pub.publish(obj)
        

    ##detaches all objects from the gripper 
    #(adds back as objects in the collision space where they are now)
    #object_collision_name is the original name for the collision object
    def detach_and_add_back_objects_attached_to_gripper(self, whicharm = 'r', object_collision_name = None):

        rospy.loginfo("detaching all objects from gripper %s and adding them back to the collision map"%whicharm)

        if object_collision_name == None:
            rospy.loginfo("need to specify the object name!  Detaching and not adding back object.")
            self.detach_all_objects_from_gripper(whicharm)
            return

        obj = AttachedCollisionObject()
        obj.object.header.stamp = rospy.Time.now()
        obj.object.header.frame_id = "base_link"
        obj.link_name = whicharm+"_gripper_r_finger_tip_link"
        obj.object.id = object_collision_name
        obj.object.operation.operation = CollisionObjectOperation.DETACH_AND_ADD_AS_OBJECT
        self.attached_object_pub.publish(obj)


    ##convert a Pose message to a 4x4 scipy matrix
    def pose_to_mat(self, pose):
        quat = [pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w]
        pos = scipy.matrix([pose.position.x, pose.position.y, pose.position.z]).T
        mat = scipy.matrix(tf.transformations.quaternion_matrix(quat))
        mat[0:3, 3] = pos
        return mat


    #convert a 4x4 scipy matrix to a Pose message
    def mat_to_pose(self, mat):
        pose = Pose()
        pose.position.x = mat[0,3]
        pose.position.y = mat[1,3]
        pose.position.z = mat[2,3]
        quat = tf.transformations.quaternion_from_matrix(mat)
        pose.orientation.x = quat[0]
        pose.orientation.y = quat[1]
        pose.orientation.z = quat[2]
        pose.orientation.w = quat[3]
        return pose


    #adds the table to the map
    def process_collision_geometry_for_table(self, firsttable, additional_tables = []):

        table_object = CollisionObject()
        table_object.operation.operation = CollisionObjectOperation.ADD
        table_object.header.frame_id = firsttable.pose.header.frame_id
        table_object.header.stamp = rospy.Time.now()

        #create a box for each table
        for table in [firsttable,]+additional_tables:
            object = Shape()
            object.type = Shape.BOX;
            object.dimensions.append(math.fabs(table.x_max-table.x_min))
            object.dimensions.append(math.fabs(table.y_max-table.y_min))
            object.dimensions.append(0.01)
            table_object.shapes.append(object)
  
        #set the origin of the table object in the middle of the firsttable
        table_mat = self.pose_to_mat(firsttable.pose.pose)
        table_offset = scipy.matrix([(firsttable.x_min + firsttable.x_max)/2.0, (firsttable.y_min + firsttable.y_max)/2.0, 0.0]).T
        table_offset_mat = scipy.matrix(scipy.identity(4))
        table_offset_mat[0:3,3] = table_offset
        table_center = table_mat * table_offset_mat
        origin_pose = self.mat_to_pose(table_center)
        table_object.poses.append(origin_pose)

        table_object.id = "table"
        self.object_in_map_pub.publish(table_object)


    ##return the next unique collision object name
    def get_next_object_name(self):
        string = "graspable_object_"+str(self.object_in_map_current_id)
        self.object_in_map_current_id += 1
        return string


    ##adds a single box to the map
    def add_collision_box(self, box_pose, box_dims, frame_id, collision_name):
        
        rospy.loginfo("adding box to collision map")
        box = CollisionObject()
        box.operation.operation = CollisionObjectOperation.ADD
        box.header.frame_id = frame_id
        box.header.stamp = rospy.Time.now()
        shape = Shape()
        shape.type = Shape.BOX
        shape.dimensions = box_dims
        box.shapes.append(shape)
        box.poses.append(box_pose)
        box.id = collision_name
        self.object_in_map_pub.publish(box)


    ##adds a cluster to the map as a bunch of small boxes
    def process_collision_geometry_for_cluster(self, cluster):

        rospy.loginfo("adding cluster with %d points to collision map"%len(cluster.points))

        many_boxes = CollisionObject()
        
        many_boxes.operation.operation = CollisionObjectOperation.ADD
        many_boxes.header = cluster.header
        many_boxes.header.stamp = rospy.Time.now()
        num_to_use = int(len(cluster.points)/100.0)
        random_indices = range(len(cluster.points))
        scipy.random.shuffle(random_indices)
        random_indices = random_indices[0:num_to_use]
        for i in range(num_to_use):
            shape = Shape()
            shape.type = Shape.BOX
            shape.dimensions = [.005]*3
            pose = Pose()
            pose.position.x = cluster.points[random_indices[i]].x
            pose.position.y = cluster.points[random_indices[i]].y
            pose.position.z = cluster.points[random_indices[i]].z
            pose.orientation = Quaternion(*[0,0,0,1])
            many_boxes.shapes.append(shape)
            many_boxes.poses.append(pose)
        
        collision_name = self.get_next_object_name()
        many_boxes.id = collision_name
        self.object_in_map_pub.publish(many_boxes)
        return collision_name


    ##remove a single collision object from the map
    def remove_collision_object(self, collision_name):

        reset_object = CollisionObject()
        reset_object.operation.operation = CollisionObjectOperation.REMOVE
        reset_object.header.frame_id = "base_link"
        reset_object.header.stamp = rospy.Time.now()
        reset_object.id = collision_name
        self.object_in_map_pub.publish(reset_object) 

File: overlay/tabletop_object_perception/tabletop_collision_map_processing/src/collision_map_interface.cpp
/*********************************************************************
*
*  Copyright (c) 2009, Willow Garage, Inc.
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*   * Neither the name of the Willow Garage nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*********************************************************************/

// Author(s): Gil Jones and Matei Ciocarlie

#include "tabletop_collision_map_processing/collision_map_interface.h"

#include <tf/transform_datatypes.h>

#include <std_srvs/Empty.h>

#include <mapping_msgs/CollisionObject.h>
#include <mapping_msgs/AttachedCollisionObject.h>

#include <object_manipulation_msgs/FindClusterBoundingBox.h>

#include <household_objects_database_msgs/DatabaseModelPose.h>
#include <household_objects_database_msgs/GetModelMesh.h>

namespace tabletop_collision_map_processing {

static const std::string RESET_COLLISION_MAP_NAME = "collision_map_self_occ_node/reset";
static const std::string MAKE_STATIC_COLLISION_MAP_ACTION_NAME = "make_static_collision_map";
static const std::string CLUSTER_BOUNDING_BOX_NAME = "find_cluster_bounding_box";

CollisionMapInterface::CollisionMapInterface() : 
  root_nh_(""),
  priv_nh_("~"),
  collision_object_current_id_(0),
  make_static_collision_map_client_(root_nh_, MAKE_STATIC_COLLISION_MAP_ACTION_NAME, true)
{
  collision_object_pub_ = root_nh_.advertise<mapping_msgs::CollisionObject>("collision_object", 10);

  attached_object_pub_ = root_nh_.advertise<mapping_msgs::AttachedCollisionObject>("attached_collision_object", 10);
  
  while ( !ros::service::waitForService(RESET_COLLISION_MAP_NAME, ros::Duration(2.0)) && root_nh_.ok() ) 
  {
    ROS_INFO("Waiting for reset collision map service to come up");
  }
  collision_map_reset_client_ = root_nh_.serviceClient<std_srvs::Empty>(RESET_COLLISION_MAP_NAME, true);
  
  while ( !ros::service::waitForService(CLUSTER_BOUNDING_BOX_NAME, ros::Duration(2.0)) && root_nh_.ok() ) 
  {
    ROS_INFO("Waiting for cluster bounding box service to come up");
  }
  cluster_bounding_box_client_ = root_nh_.serviceClient<object_manipulation_msgs::FindClusterBoundingBox>
    (CLUSTER_BOUNDING_BOX_NAME, true);

  while(!make_static_collision_map_client_.waitForServer(ros::Duration(2.0))
	&& root_nh_.ok()){
    ROS_INFO("Waiting for the make static collision map action to come up");
  }
}

void CollisionMapInterface::resetCollisionModels()
{
  mapping_msgs::CollisionObject reset_object;
  reset_object.operation.operation = mapping_msgs::CollisionObjectOperation::REMOVE;
  reset_object.header.frame_id = "base_link";
  reset_object.header.stamp = ros::Time::now();
  reset_object.id = "all";
  collision_object_pub_.publish(reset_object);
  collision_object_current_id_ = 0;
}

void CollisionMapInterface::resetAttachedModels()
{
  mapping_msgs::CollisionObject reset_object;
  reset_object.operation.operation = mapping_msgs::CollisionObjectOperation::REMOVE;
  reset_object.header.frame_id = "base_link";
  reset_object.header.stamp = ros::Time::now();
  reset_object.id = "all";
  mapping_msgs::AttachedCollisionObject reset_attached_objects;
  reset_attached_objects.object.header.frame_id = "base_link";
  reset_attached_objects.object.header.stamp = ros::Time::now();
  reset_attached_objects.link_name = "all";
  reset_attached_objects.object = reset_object;
  attached_object_pub_.publish(reset_attached_objects);
}

void CollisionMapInterface::resetStaticMap() 
{
  std_srvs::Empty srv; 
  if (!collision_map_reset_client_.call(srv))
  {
    ROS_ERROR("Collision map reset call failed");
    throw CollisionMapException("reset failed");
  }
}

std::string CollisionMapInterface::getNextObjectName()
{
  std::ostringstream iss;
  iss << collision_object_current_id_++;
  return "graspable_object_" + iss.str();
}

void CollisionMapInterface::takeStaticMap() 
{  
  collision_environment_msgs::MakeStaticCollisionMapGoal static_map_goal;
  static_map_goal.cloud_source = "camera/depth/points";
  static_map_goal.number_of_clouds = 2;
  
  make_static_collision_map_client_.sendGoal(static_map_goal);

  if ( !make_static_collision_map_client_.waitForResult(ros::Duration(30.0)) )
  {
    ROS_ERROR("Collision map was not formed in allowed time");
    throw CollisionMapException("static make was not formed in allowed time");
  }

  if(make_static_collision_map_client_.getState() != actionlib::SimpleClientGoalState::SUCCEEDED) 
  {
    ROS_ERROR("Some non-success state was reached for static collision map.  Proceed with caution");
    throw CollisionMapException(" static collision map failed");
  } 
}

void CollisionMapInterface::processCollisionGeometryForTable(const tabletop_object_detector::Table &table,
							     std::string table_collision_name) 
{
  mapping_msgs::CollisionObject table_object;
  table_object.operation.operation = mapping_msgs::CollisionObjectOperation::ADD;
  table_object.header.frame_id = table.pose.header.frame_id;
  table_object.header.stamp = ros::Time::now();
  table_object.shapes.resize(1);
  table_object.shapes[0].type = geometric_shapes_msgs::Shape::BOX;
  table_object.shapes[0].dimensions.resize(3);
  table_object.shapes[0].dimensions[0] = fabs(table.x_max-table.x_min);
  table_object.shapes[0].dimensions[1] = fabs(table.y_max-table.y_min);
  table_object.shapes[0].dimensions[2] = 0.01;

  //set the origin of the table object in the middle of the table
  tf::Transform table_trans;
  tf::poseMsgToTF(table.pose.pose, table_trans);
  tf::Transform table_translation;
  table_translation.setIdentity();
  table_translation.setOrigin( btVector3( (table.x_min + table.x_max)/2.0, (table.y_min + table.y_max)/2.0, 0.0) );
  table_trans = table_trans * table_translation;
  table_object.poses.resize(1);
  tf::poseTFToMsg(table_trans, table_object.poses[0]);

  table_object.id = table_collision_name;
  collision_object_pub_.publish(table_object);
}

void CollisionMapInterface::processCollisionGeometryForObject
  (const household_objects_database_msgs::DatabaseModelPose &model_pose,
   std::string &collision_name)
{
  mapping_msgs::CollisionObject collision_object;
  collision_object.shapes.resize(1);

  if (!getMeshFromDatabasePose(model_pose, collision_object.shapes[0]))
  {
    throw CollisionMapException("Loading mesh for database object failed");
  }
  collision_object.header.frame_id = model_pose.pose.header.frame_id;
  collision_object.header.stamp = ros::Time::now();
  collision_object.poses.push_back(model_pose.pose.pose);
  
  collision_object.shapes[0].type = geometric_shapes_msgs::Shape::MESH;
  collision_object.operation.operation = mapping_msgs::CollisionObjectOperation::ADD;
  collision_name = getNextObjectName();
  collision_object.id = collision_name;
  collision_object_pub_.publish(collision_object);
}

void 
CollisionMapInterface::processCollisionGeometryForBoundingBox(const object_manipulation_msgs::ClusterBoundingBox &box, 
							      std::string &collision_name)
{
  ROS_INFO("Adding bounding box with dimensions %f %f %f to collision map", 
	   box.dimensions.x, box.dimensions.y, box.dimensions.z);

  mapping_msgs::CollisionObject collision_object;
  collision_object.operation.operation = mapping_msgs::CollisionObjectOperation::ADD;

  collision_name = getNextObjectName();
  collision_object.id = collision_name;

  collision_object.header.frame_id = box.pose_stamped.header.frame_id;
  collision_object.header.stamp = ros::Time::now();
  
  geometric_shapes_msgs::Shape shape;
  shape.type = geometric_shapes_msgs::Shape::BOX;
  shape.dimensions.resize(3);
  shape.dimensions[0] = box.dimensions.x;
  shape.dimensions[1] = box.dimensions.y;
  shape.dimensions[2] = box.dimensions.z;
  collision_object.shapes.push_back(shape);
  collision_object.poses.push_back(box.pose_stamped.pose);

  collision_object_pub_.publish(collision_object);
}

void CollisionMapInterface::processCollisionGeometryForCluster(const sensor_msgs::PointCloud &cluster,
							       std::string &collision_name)
{
  ROS_INFO("Adding cluster with %u points to collision map", (unsigned int)cluster.points.size()); 

  mapping_msgs::CollisionObject many_boxes;
  many_boxes.operation.operation = mapping_msgs::CollisionObjectOperation::ADD;
  many_boxes.header = cluster.header;
  many_boxes.header.stamp = ros::Time::now();
  unsigned int num_to_use = (unsigned int)(cluster.points.size()/100.0);
  many_boxes.shapes.resize(num_to_use);
  many_boxes.poses.resize(num_to_use);
  for(unsigned int i = 0; i < num_to_use; i++) {
    geometric_shapes_msgs::Shape shape;
    shape.type = geometric_shapes_msgs::Shape::BOX;
    shape.dimensions.resize(3);
    shape.dimensions[0] = .005;
    shape.dimensions[1] = .005;
    shape.dimensions[2] = .005;
    many_boxes.shapes[i]=shape;
    geometry_msgs::Pose pose;
    pose.position.x = cluster.points[i*100].x;
    pose.position.y = cluster.points[i*100].y;
    pose.position.z = cluster.points[i*100].z;
    pose.orientation.x = 0;
    pose.orientation.y = 0;
    pose.orientation.z = 0;
    pose.orientation.w = 1;
    many_boxes.poses[i] = pose;
  }

  collision_name = getNextObjectName();
  //use this name for the collision map
  many_boxes.id = collision_name;
  collision_object_pub_.publish(many_boxes);
}

bool CollisionMapInterface::getMeshFromDatabasePose(const household_objects_database_msgs::DatabaseModelPose &model_pose,
						    geometric_shapes_msgs::Shape& mesh) 
{
  static bool service_initialized = false;
  if (!service_initialized)
  {    
    std::string get_model_mesh_srv_name;
    priv_nh_.param<std::string>("get_model_mesh_srv", get_model_mesh_srv_name, "get_model_mesh_srv");
    while ( !ros::service::waitForService(get_model_mesh_srv_name, ros::Duration(2.0)) && root_nh_.ok() ) 
    {
      ROS_INFO("Waiting for %s service to come up", get_model_mesh_srv_name.c_str());
    }
    if (!root_nh_.ok()) exit(0);
    get_model_mesh_srv_ = root_nh_.serviceClient<household_objects_database_msgs::GetModelMesh>
      (get_model_mesh_srv_name, true);
    service_initialized = true;
  }
  household_objects_database_msgs::GetModelMesh get_mesh;
  get_mesh.request.model_id = model_pose.model_id;
  if ( !get_model_mesh_srv_.call(get_mesh) || 
       get_mesh.response.return_code.code != get_mesh.response.return_code.SUCCESS )
  {
    return false;
  }
  mesh = get_mesh.response.mesh;
  return true;
}

void CollisionMapInterface::getClusterBoundingBox(const sensor_msgs::PointCloud &cluster,
						  geometry_msgs::PoseStamped &pose_stamped, 
						  geometry_msgs::Vector3 &dimensions)
{
  object_manipulation_msgs::FindClusterBoundingBox srv;
  srv.request.cluster = cluster;
  if (!cluster_bounding_box_client_.call(srv.request, srv.response))
  {
    ROS_ERROR("Failed to call cluster bounding box client");
    throw CollisionMapException("Failed to call cluster bounding box client");
  }
  pose_stamped = srv.response.pose;
  dimensions = srv.response.box_dims;
  if (dimensions.x == 0.0 && dimensions.y == 0.0 && dimensions.z == 0.0)
  {
    ROS_ERROR("Cluster bounding box client returned an error (0.0 bounding box)");
    throw CollisionMapException("Bounding box computation failed");  
  }
}

/*! Assumes the bounding box is in the same frame as the table */
bool CollisionMapInterface::extendBoundingBoxZToTable(const tabletop_object_detector::Table &table,
                                                      geometry_msgs::PoseStamped &pose_stamped, 
                                                      geometry_msgs::Vector3 &dimensions)
{
  if (table.pose.header.frame_id != pose_stamped.header.frame_id)
  {
    ROS_INFO("cannot extend bbox to table, they are not in the same frame");
    return false;
  }

  //get the distance from the box to the table
  btVector3 table_location(table.pose.pose.position.x,
                           table.pose.pose.position.y,
                           table.pose.pose.position.z);
  btVector3 bbox_location(pose_stamped.pose.position.x, 
                          pose_stamped.pose.position.y, 
                          pose_stamped.pose.position.z);
  btVector3 table_to_bbox = bbox_location - table_location;
  btTransform table_trans;
  tf::poseMsgToTF(table.pose.pose, table_trans);
  btVector3 table_z = table_trans.getBasis().getColumn(2);
  ROS_INFO("Table z: %f %f %f in frame %s", table_z.getX(), table_z.getY(), table_z.getZ(), table.pose.header.frame_id.c_str());
  double box_to_table_distance = table_to_bbox.dot(table_z);
  ROS_INFO("Table distance: %f", box_to_table_distance);

  //now we actually make the assumptions that the z axes are close to each other
  //as doing an actual computation was just too hard
  btTransform bbox_trans;
  tf::poseMsgToTF(pose_stamped.pose, bbox_trans);
  btVector3 bbox_z = bbox_trans.getBasis().getColumn(2);
  if (fabs(bbox_z.dot(table_z)) < 0.9848) //10 degrees
  { 
    ROS_INFO("cannot extend bbox to table; z axes do not coincide");
    return false;
  }
  
  double z_to_table = box_to_table_distance;
  ROS_INFO("z_to_table: %f", z_to_table);
  if ( z_to_table < 0.005 || z_to_table > 0.3)
  {
    ROS_INFO("cannot extend bbox to table; getting z equal to %f", z_to_table);
    return false;
  }
  ROS_INFO("Old z: %f", dimensions.z);
  double new_z = dimensions.z/2.0 + z_to_table;
  ROS_INFO("New z: %f", new_z);
  pose_stamped.pose.position.z = pose_stamped.pose.position.z + (dimensions.z/2.0) - (new_z/2.0);
  dimensions.z = new_z;
  return true;
}


} //namespace tabletop_collision_map_processing
